src/dsl.jl

```diff
@@ -688,25 +692,41 @@ function read_equations_options(options, variables_declared)
             error("Malformed equation: \"$eq\". Equation's left hand and right hand sides should be separated by a \"~\".")
         end
 
-        # Checks if the equation have the format D(X) ~ ... (where X is a symbol). This means that the
-        # default differential has been used. X is added as a declared variable to the system, and
-        # we make a note that a differential D = Differential(iv) should be made as well.
+        # Check if the defualt differential is used. If so, declare it and its variable. 
         lhs = eq.args[2]
-        # if lhs: is an expression. Is a function call. The function's name is D. Calls a single symbol.
-        if (lhs isa Expr) && (lhs.head == :call) && (lhs.args[1] == :D) &&
-           (lhs.args[2] isa Symbol)
-            diff_var = lhs.args[2]
-            if in(diff_var, forbidden_symbols_error)
+        ((lhs isa Symbol) || (lhs isa Number)) && continue
+
+        diff_vars = Vector{Symbol}()
+        if uses_default_differential!(lhs, diff_vars)
+            add_default_diff = true
+            if any(var -> var in forbidden_symbols_error, diff_vars)
+                bad_var = findfirst(var -> var in forbidden_symbols_error, diff_vars)
                 error("A forbidden symbol ($(diff_var)) was used as an variable in this differential equation: $eq")
             end
-            add_default_diff = true
-            in(diff_var, variables_declared) || push!(vars_extracted, diff_var)
         end
+        push!(vars_extracted, diff_vars...)
     end
+    vars_extracted = filter(var -> var ∉ variables_declared, vars_extracted)
 
     return vars_extracted, add_default_diff, equations
 end
 
+# Checks if the equation have the format f(D(X)) ~ ... (where X is a symbol). This means that the
+# default differential has been used. X is added as a declared variable to the system, and
+# we make a note that a differential D = Differential(iv) should be made as well.
+function uses_default_differential!(lhs::Expr, diff_vars)
+    if (lhs isa Expr) && (lhs.head == :call) && (lhs.args[1] == :D) &&
+       (lhs.args[2] isa Symbol)
+        push!(diff_vars, lhs.args[2])
+    end
+
+    for arg in lhs.args
+        isa(arg, Expr) && uses_default_differential!(arg, diff_vars)
+    end
+
+    isempty(diff_vars) ? (return false) : (return true)
+end
+
```

Comment:
@vyudu are these changes needed for the function bug fix? If not, can we move them to a followup PR to expand the supported LHS's in ODEs? Let's keep this focused on just fixing the bug with the current setup.